v0/destinations/salesforce/utils.js

```diff
@@ -0,0 +1,119 @@
+  SF_TOKEN_REQUEST_URL,
+  DESTINATION
+} = require("./config");
+
+const ACCESS_TOKEN_CACHE = new Cache(ACCESS_TOKEN_CACHE_TTL);
+
+// Utility method to construct the header to be used for SFDC API calls
+// The "Authorization: Bearer <token>" header element needs to be passed for
+// authentication for all SFDC REST API calls
+const getAccessToken = async destination => {
+  const accessTokenKey = destination.ID;
+
+  return ACCESS_TOKEN_CACHE.get(accessTokenKey, async () => {
+    let SF_TOKEN_URL;
+    if (destination.Config.sandbox) {
+      SF_TOKEN_URL = SF_TOKEN_REQUEST_URL_SANDBOX;
+    } else {
+      SF_TOKEN_URL = SF_TOKEN_REQUEST_URL;
+    }
+    const authUrl = `${SF_TOKEN_URL}?username=${
+      destination.Config.userName
+    }&password=${encodeURIComponent(
+      destination.Config.password
+    )}${encodeURIComponent(destination.Config.initialAccessToken)}&client_id=${
+      destination.Config.consumerKey
+    }&client_secret=${destination.Config.consumerSecret}&grant_type=password`;
+    const salesforceAuthorisationData = await httpPOST(authUrl, {});
+    const salesforceAxiosResponse = processAxiosResponse(
+      salesforceAuthorisationData
+    );
+    // If the request fails, throwing error.
+    if (salesforceAuthorisationData.success === false) {
+      const { error } = salesforceAuthorisationData.response.response.data;
+      throw new ApiError(
+        `${DESTINATION} Request Failed: access token could not be generated due to ${error}`,
+        400,
+        {
+          scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
+          meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.META.ABORTABLE
+        },
+        salesforceAxiosResponse.response,
+        undefined,
+        DESTINATION
+      );
+    }
+    const token = salesforceAuthorisationData.response.data;
+    return {
+      token: `Bearer ${token.access_token}`,
+      instanceUrl: token.instance_url
+    };
+  });
+};
+
+const salesforceResponseHandler = (
+  destResponse,
+  sourceMessage,
+  stage,
+  authKey
+) => {
+  const { status, response } = destResponse;
+
+  // if the response from destination is not a success case build an explicit error
+  if (!isHttpStatusSuccess(status)) {
+    if (status === 401 && authKey) {
+      if (
+        response &&
+        Array.isArray(response) &&
+        response[0].errorCode === "INVALID_SESSION_ID"
+      ) {
+        // checking for invalid/expired token errors and evicting cache in that case
+        // rudderJobMetadata contains some destination info which is being used to evict the cache
+        ACCESS_TOKEN_CACHE.del(authKey);
+        throw new ApiError(
+          `${DESTINATION} Request Failed - due to ${response[0].message}, (Retryable).${sourceMessage}`,
+          500,
+          {
+            scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
+            meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.META.RETRYABLE,
+            stage
+          },
+          sourceMessage,
```

Comment:
The argument after `statTags` in APIError class indicates original destination response that was sent. I think this should be either `response` or `destResponse`. I'd put `destResponse` here.